File: apple/IllUsionVPN/Models/Session.swift
import Foundation

/// WireGuard-сессия, выдаваемая backend (`POST /api/session`).
struct Session: Codable {
    struct Interface: Codable {
        let address: [String]
        let dns: [String]
        let mtu: Int
    }

    struct Peer: Codable {
        let publicKey: String
        let endpoint: String
        let allowedIPs: [String]
        let persistentKeepalive: Int?
    }

    let sessionId: String
    let multihop: Bool?
    let interface: Interface
    let peers: [Peer]
    let expiresAt: Double

    /// Сборка конфигурации в формате wg-quick (`[Interface]` / `[Peer]`).
    /// `privateKey` генерируется и хранится на устройстве, не уходит на сервер.
    func wgQuickConfig(privateKey: String) -> String {
        var lines: [String] = []
        lines.append("[Interface]")
        lines.append("PrivateKey = \(privateKey)")
        lines.append("Address = \(interface.address.joined(separator: ", "))")
        lines.append("DNS = \(interface.dns.joined(separator: ", "))")
        lines.append("MTU = \(interface.mtu)")
        for peer in peers {
            lines.append("")
            lines.append("[Peer]")
            lines.append("PublicKey = \(peer.publicKey)")
            lines.append("Endpoint = \(peer.endpoint)")
            lines.append("AllowedIPs = \(peer.allowedIPs.joined(separator: ", "))")
            if let keepalive = peer.persistentKeepalive {
                lines.append("PersistentKeepalive = \(keepalive)")
            }
        }
        return lines.joined(separator: "\n")
    }
}

File: windows/IllUsionVPN/Models/Session.cs
using System.Collections.Generic;
using System.Text;
using System.Text.Json.Serialization;

namespace IllUsionVPN.Models;

public sealed class Session
{
    public sealed class InterfaceConfig
    {
        [JsonPropertyName("address")] public List<string> Address { get; set; } = new();
        [JsonPropertyName("dns")] public List<string> Dns { get; set; } = new();
        [JsonPropertyName("mtu")] public int Mtu { get; set; } = 1420;
    }

    public sealed class Peer
    {
        [JsonPropertyName("publicKey")] public string PublicKey { get; set; } = "";
        [JsonPropertyName("endpoint")] public string Endpoint { get; set; } = "";
        [JsonPropertyName("allowedIPs")] public List<string> AllowedIPs { get; set; } = new();
        [JsonPropertyName("persistentKeepalive")] public int? PersistentKeepalive { get; set; }
    }

    [JsonPropertyName("sessionId")] public string SessionId { get; set; } = "";
    [JsonPropertyName("multihop")] public bool? Multihop { get; set; }
    [JsonPropertyName("interface")] public InterfaceConfig Interface { get; set; } = new();
    [JsonPropertyName("peers")] public List<Peer> Peers { get; set; } = new();
    [JsonPropertyName("expiresAt")] public double ExpiresAt { get; set; }

    /// <summary>Собирает конфигурацию в формате wg-quick для tunnel.dll.</summary>
    public string ToWgQuickConfig(string privateKey)
    {
        var sb = new StringBuilder();
        sb.AppendLine("[Interface]");
        sb.AppendLine($"PrivateKey = {privateKey}");
        sb.AppendLine($"Address = {string.Join(", ", Interface.Address)}");
        sb.AppendLine($"DNS = {string.Join(", ", Interface.Dns)}");
        sb.AppendLine($"MTU = {Interface.Mtu}");
        foreach (var peer in Peers)
        {
            sb.AppendLine();
            sb.AppendLine("[Peer]");
            sb.AppendLine($"PublicKey = {peer.PublicKey}");
            sb.AppendLine($"Endpoint = {peer.Endpoint}");
            sb.AppendLine($"AllowedIPs = {string.Join(", ", peer.AllowedIPs)}");
            if (peer.PersistentKeepalive is int k)
                sb.AppendLine($"PersistentKeepalive = {k}");
        }
        return sb.ToString();
    }
}
